### PR TITLE
[Fix] Improve bson usage

### PR DESCRIPTION
**Pull request by stevensJourney**
# Description

Some issues were raised with the recent addition of the `bson` package for WebSocket streaming.

## Vite warnings
A warning for `crypto.randomBytes` being used would be shown in Vite applications even if the WebSockets mode was not used.

The warning is now removed due to using a dynamic import of `bson` in the Web SDK. The Vite config `resolve` setting is now no longer required.
 
## Webpack example

The example Webpack and NextJS demos was broken due to bundling issues. BSON requires certain dependencies to be externalised. These have been added to the example config. These are not required for NextJS configs.

## React Native

In some React Native environments the wrong import might be used, causing top level await issues. The React Native import has been fixed according to [these](https://github.com/mongodb/js-bson?tab=readme-ov-file#react-native) docs.

# Testing

This was tested on all the demos with HTTP streaming and Web socket modes enabled.

NextJS demo working for demonstration:
image

### CODE REVIEW

File: demos/example-electron/vite.renderer.config.ts [hunk 1]
```diff
@@ -30,7 +30,6 @@ export default defineConfig((env) => {
         '@powersync/web > js-logger',
         '@powersync/web > lodash/throttle',
         '@powersync/web > can-ndjson-stream',
-        '@powersync/web > bson',
```
Comment (stevensJourney) [inline demos/example-electron/vite.renderer.config.ts:33]: This is no longer needed in our Vite configs

File: demos/react-supabase-todolist/package.json [hunk 2]
```diff
@@ -11,12 +11,15 @@
   "dependencies": {
     "@powersync/react": "workspace:*",
     "@powersync/web": "workspace:*",
+    "@emotion/react": "11.11.4",
```
Comment (stevensJourney) [inline demos/react-supabase-todolist/package.json:14]: Added these as it was necessary when copying this project our to a standalone demo folder.

File: packages/common/src/client/sync/stream/AbstractRemote.ts [hunk 3]
```diff
@@ -171,13 +173,20 @@ export abstract class AbstractRemote {
...
+  abstract getBSON(): Promise<BSONImplementation>;
+
   /**
    * Connects to the sync/stream websocket endpoint
    */
   async socketStream(options: SyncStreamOptions): Promise<DataStream<StreamingSyncLine>> {
     const { path } = options;
     const request = await this.buildRequest(path);
 
+    const BSON = await this.getBSON();
```
Comment (DominicGBauer) [inline packages/common/src/client/sync/stream/AbstractRemote.ts:188]: can this be lower case so that it doesn't get confused with BSON type?

Comment (stevensJourney) [reply packages/common/src/client/sync/stream/AbstractRemote.ts]: Can do that. It's now lowercase.

**Review by DominicGBauer — APPROVED** [review]

**Review by Chriztiaan — APPROVED** [review]

**Pull request merged**